### PR TITLE
Several PRs combined into a single PR.

### PR DESCRIPTION
**Pull request by olafurpg**
This is the combination of

- #986 
- #984 
- #985 
- #989 
- #992
- this original PR, which fixed several issues with sugars.

All of the PRs did something small that caused merge conflicts with each other. It's much simpler to manage them together.

### CODE REVIEW

File: scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala [hunk 1]
```diff
@@ -264,13 +265,15 @@ trait AttributesOps { self: DatabaseOps =>
               gtree match {
                 case gview: g.ApplyImplicitView =>
                   val pos = gtree.pos.toMeta
-                  val syntax = gview.fun + "(*)"
+                  val syntax = g.showCode(gview.fun) + "(*)"
```
Comment (xeno-by) [inline scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala]: What's the difference between `toString` and `showCode` here? I'm very wary of `showCode`, because it's known to sometimes swallow snippets of code that it considers irrelevant.

File: scalahost/nsc/src/test/scala/scala/meta/tests/scalahost/mirrors/SemanticSuite.scala [hunk 2]
```diff
@@ -365,4 +365,43 @@ class SemanticSuite extends DatabaseSuite(SemanticdbMode.Slim) {
...
+  sugars(
+    """|object b {
+       |  class F
+       |  implicit val ordering: Ordering[F] = ???
+       |  val x: Ordered[F] = new F
+       |}
+    """.stripMargin,
+    """|[86..91) scala.math.Ordered.orderingToOrdered[b.F](*)(b.this.ordering)
+       |""".stripMargin
+  )
```
Comment (xeno-by) [inline scalahost/nsc/src/test/scala/scala/meta/tests/scalahost/mirrors/SemanticSuite.scala]: Please rename the enclosing objects in the tests once we merge #907.

File: scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala (hunk 1 above)
Comment (olafurpg) [inline scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala]: Here's an example failing test from rebase on master

```
[info]   -[273..275) h.X.cvt[Int](*)(h.C.int)
[info]   +[273..275) h.this.X.cvt[Int](*)(h.this.C.int)
```
The second one (.toString) would not compile if inserted into a source file while the first one (showCode) would compile.

**Comment by olafurpg** [comment]
CI on appveyor is already proving itself valuable, this surfaced in the new expect tests

```
[info]   =======
[info]   => Diff
[info]   =======
[info]   -scalahost\integration\src\main\scala\example\Example.scala
[info]   +scalahost/integration/src/main/scala/example/Example.scala
[info]    ---------------------------------------------------------- (SemanticdbExpectSuite.scala:14)
[info] ScalaTest
```

**Comment by olafurpg** [comment]
No new features or changes were introduced from yesterday. I'm gonna merge this now to unblock further changes today.

File: scalahost/nsc/src/main/scala/scala/meta/internal/scalahost/ScalahostPipeline.scala [hunk 3]
```diff
@@ -14,6 +14,7 @@ import scala.meta.internal.semantic.DatabaseOps
 import scala.meta.internal.semantic.{vfs => v}
 import scala.meta.internal.semantic.{schema => s}
 import scala.tools.nsc.doc.ScaladocGlobal
+import scala.meta.internal.semantic.schema.Message.Severity
```
Comment (xeno-by) [inline scalahost/nsc/src/main/scala/scala/meta/internal/scalahost/ScalahostPipeline.scala]: Looks like an unused import.

File: scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala [hunk 4]
```diff
@@ -325,12 +328,16 @@ trait AttributesOps { self: DatabaseOps =>
                   val morePrecisePos = fun.pos.withStart(fun.pos.end).toMeta
                   val syntax = "[" + targs.mkString(", ") + "]"
                   success(morePrecisePos, syntax)
+                case g.Apply(select @ g.Select(qual, nme), _) if isSyntheticName(select) =>
+                  val pos = qual.pos.withStart(qual.pos.end).toMeta
+                  success(pos, s".${nme.decoded}")
```
Comment (xeno-by) [inline scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala]: Why do we need `decoded` here?

File: scalahost/nsc/src/main/scala/scala/meta/internal/semantic/ConfigOps.scala [hunk 5]
```diff
@@ -5,13 +5,32 @@ import scala.meta.internal.io.PathIO
...
+  def default = ScalahostConfig(PathIO.workingDirectory, SemanticdbMode.Fat, FailureMode.Warning)
+}
+
+sealed abstract class FailureMode {
+  def name: String = toString.toLowerCase
+}
+object FailureMode {
+  def unapply(arg: String): Option[FailureMode] = {
+    val lower = arg.toLowerCase()
+    all.find(_.name == lower)
```
Comment (xeno-by) [inline scalahost/nsc/src/main/scala/scala/meta/internal/semantic/ConfigOps.scala]: This implementation looks inconsistent with `SemanticdbMode.unapply`.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala [hunk 6]
```diff
@@ -2,18 +2,102 @@ package scala.meta
...
+        val abspath = targetroot.resolve(relpath).toNIO
+        Files.createDirectories(abspath.getParent)
+        val fos = Files.newOutputStream(abspath)
+        try {
+          s.Database(List(entry)).writeTo(fos)
+        } finally {
+          fos.close()
+        }
+      }
+    }
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala]: I think we should either consistently get rid of vfs or consistently use it in both directions.

Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala]: Let's first discuss the idea behind merging `toVfs` and `save`. Is that to save memory spent on precaching arrays of bytes? If yes, why not make `bytes` lazy?

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Database.scala [hunk 7]
```diff
@@ -19,17 +18,7 @@ object Database {
 @data
 class Database(entries: List[Entry]) {
   def toSchema: s.Database = {
-    val sentries = entries.map(ventry => s.Attributes.parseFrom(ventry.bytes))
+    val sentries = entries.flatMap(ventry => s.Database.parseFrom(ventry.inputStream).entries)
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Database.scala]: Am I right in thinking that the change from `.bytes` to `.inputStream` here is done for optimization purposes?

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala [hunk 8]
```diff
@@ -1,28 +1,26 @@
 package scala.meta.internal.semantic.vfs
 
 import scala.meta.internal.io.PathIO.fileSeparator
+import scala.meta.io.AbsolutePath
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: Looks like an unused import.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala [hunk 9]
```diff
@@ -1,28 +1,26 @@
...
+  private val semanticDbPrefix: RelativePath = RelativePath("META-INF").resolve("semanticdb")
   private val semanticDbSuffix = ".semanticdb"
   private val scalaPrefix = ""
   private val scalaSuffix = ".scala"
 
   def isSemanticdb(path: RelativePath): Boolean = {
-    path.toString.startsWith(semanticDbPrefix) &&
-    path.toString.endsWith(semanticDbSuffix)
+    path.toNIO.startsWith(semanticDbPrefix.toNIO) &&
+    path.path.getFileName.toString.endsWith(semanticDbSuffix)
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: `path.path` reads weirdly. Can we change it to `path.value`?

Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: Also, I'm not sure whether `path.path.getFileName.toString` constitutes an improvement with respect to `path.toString`. Can you comment on this?

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala [hunk 10]
```diff
@@ -1,28 +1,26 @@
...
   def semanticdbToScala(path: RelativePath): RelativePath = {
     require(isSemanticdb(path))
-    RelativePath(
-      scalaPrefix +
-        path.toString
-          .stripPrefix(semanticDbPrefix)
-          .stripSuffix(semanticDbSuffix) +
-        scalaSuffix)
+    RelativePath(path.toNIO
+      .relativize(path.toNIO)
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: `path.relativize(path)` reads like a typo. Have I misunderstood something?

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala [hunk 11]
```diff
@@ -32,11 +30,6 @@ object Paths {
 
   def scalaToSemanticdb(path: RelativePath): RelativePath = {
     require(isScala(path))
-    RelativePath(
-      semanticDbPrefix +
-        path.toString
-          .stripPrefix(scalaPrefix)
-          .stripSuffix(scalaSuffix) +
-        semanticDbSuffix)
+    semanticDbPrefix.resolve(path.resolveSibling(_.stripSuffix(scalaSuffix) + semanticDbSuffix))
```
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: I find the refactoring of logic in `Paths` hard to follow. Let's discuss a formulation that would work for both of us. What did you find unnice about the previous implementation?

File: scalameta/tests/jvm/src/test/resources/semanticdb.expect [hunk 12]
```diff
@@ -0,0 +1,26 @@
+scalahost/integration/src/main/scala/example/Example.scala
+----------------------------------------------------------
+Dialect:
+Scala212
```
Comment (xeno-by) [inline scalameta/tests/jvm/src/test/resources/semanticdb.expect]: How does this work? Do we force expect tests to Scala 2.12 only?

File: scalahost/nsc/src/main/scala/scala/meta/internal/scalahost/ScalahostPipeline.scala (hunk 3 above)
Comment (olafurpg) [inline scalahost/nsc/src/main/scala/scala/meta/internal/scalahost/ScalahostPipeline.scala]: I've removed all unused imports in the build https://github.com/scalameta/scalameta/pull/1004

File: scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala (hunk 4 above)
Comment (olafurpg) [inline scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala]: I thought that was the way to get the name of a symbol, is there a better method?

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala (hunk 6 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala]: With the previous vfs design, we had `Array[Byte]` + `s.Database` + `m.Database` loaded into memory for all .semanticdb files at some point. This is prohibitively expensive when running large-scale analysis. We want to avoid materializing an `Array[Byte]` entirely. Protos can be loaded in a streaming fashion, tag by tag, and that's what this refactoring did.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Database.scala (hunk 7 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Database.scala]: Correct.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala (hunk 8 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: Removed in https://github.com/scalameta/scalameta/pull/1004

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala (hunk 11 above)
Comment (olafurpg) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: I've refactored the logic a bit to make it cleaner. I agree my refactoring was a bit messy. I did not like the previous implementation since it was purely string based. A path is more a `List[String]` than a `String`, so when changing the extension we should try to use `resolveSibling` and when checking the extension we should check the extension against the filename instead of `Path.toString.endsWith`.

File: scalameta/tests/jvm/src/test/resources/semanticdb.expect (hunk 12 above)
Comment (olafurpg) [inline scalameta/tests/jvm/src/test/resources/semanticdb.expect]: That's fine for me by now. We can normalize the dialect to something else if you prefer. You won't have to fix this file manually though, just run the SaveExpect... main and commit.

**Comment by olafurpg** [comment]
Thank you for the review @xeno-by. I apologize for merging so quickly. I wanted to avoid merge conflicts with all the recent activity. I've addressed your feedback in #1005

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala (hunk 8 above)
Comment (ShaneDelmore) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: Can we just enable -Ywarn-unused:import to prevent humans having to catch this? If it makes the local development cycle awkward you can either disable that flag locally, or we could modify build to enable the flag in CI runs.

File: scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala (hunk 4 above)
Comment (xeno-by) [inline scalahost/nsc/src/main/scala/scala/meta/internal/semantic/AttributesOps.scala]: I mixed this up with the macro API, where you have to do `decodedName.toString`.

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala (hunk 8 above)
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: @olafurpg Can we implement an unused import check on CI via scalafix?

File: scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala (hunk 11 above)
Comment (xeno-by) [inline scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/vfs/Paths.scala]: On a number of occasions, we have seen that moving from `String` to `*Path` brings correctness benefits. In general, I'm in favor of refactorings like that, as long as they are reasonably concise. I had problems with the first take on this, but I do like the new version. Thank you!

File: scalameta/tests/jvm/src/test/resources/semanticdb.expect (hunk 12 above)
Comment (xeno-by) [inline scalameta/tests/jvm/src/test/resources/semanticdb.expect]: 2.12 was released quite a while ago, so it's only fair that we use it as default. Also see https://github.com/scalameta/scalameta/issues/1011.

**Comment by xeno-by** [comment]
@olafurpg It's alright. I should've communicated better by explicitly requesting changes on this PR, and I appreciate that you followed up with a new pull request so quickly. 

I suggest that in the future we don't merge without a green review. I haven't been rigorous about this in the past, but I will do better in the future. 

I really like the fact that in Scalameta we don't have processes that prevent us from moving fast. In my opinion, occasional breakages are fine, as long as putting out fires doesn't become our lifestyle. I think that the new policy is lightweight enough for us to remain comfortably productive. Wdyt?
